io.openems.edge.controller.pvinverter.reversepowerrelay/src/io/openems/edge/controller/pvinverter/reversepowerrelay/ReversePowerRelayImpl.java

```diff
@@ -0,0 +1,169 @@
+		implements ReversePowerRelay, Controller, OpenemsComponent {
+
+	private final Logger log = LoggerFactory.getLogger(ReversePowerRelayImpl.class);
+
+	@Reference
+	private ComponentManager componentManager;
+
+	private String pvInverterId;
+
+	private int powerLimit30Percent = 0;
+	private int powerLimit60Percent = 0;
+
+	private ChannelAddress inputChannelAddress0Percent = null;
+	private ChannelAddress inputChannelAddress30Percent = null;
+	private ChannelAddress inputChannelAddress60Percent = null;
+	private ChannelAddress inputChannelAddress100Percent = null;
+
+	private ManagedSymmetricPvInverter pvInverter;
+
+	public ReversePowerRelayImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				Controller.ChannelId.values(), //
+				ReversePowerRelay.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.pvInverterId = config.pvInverter_id();
+
+		this.powerLimit30Percent = config.powerLimit30();
+		this.powerLimit60Percent = config.powerLimit60();
+
+		try {
+			this.inputChannelAddress0Percent = ChannelAddress.fromString(config.inputChannelAddress0Percent());
+			this.inputChannelAddress30Percent = ChannelAddress.fromString(config.inputChannelAddress30Percent());
+
+			this.inputChannelAddress60Percent = ChannelAddress.fromString(config.inputChannelAddress60Percent());
+			this.inputChannelAddress100Percent = ChannelAddress.fromString(config.inputChannelAddress100Percent());
+
+		} catch (OpenemsNamedException e) {
+			this.log.error("Error parsing channel addresses", e);
+		}
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		// Reset limit
+		ManagedSymmetricPvInverter pvInverter;
+		try {
+			pvInverter = this.componentManager.getComponent(this.pvInverterId);
+			pvInverter.setActivePowerLimit(null);
+		} catch (OpenemsNamedException e) {
+			this.logError(this.log, e.getMessage());
+		}
+		super.deactivate();
+	}
+
+	private void setPvLimit(Integer powerLimit) {
+
+		try {
+			this.pvInverter = this.componentManager.getComponent(this.pvInverterId);
+			if (this.pvInverter != null) {
+				this.pvInverter.setActivePowerLimit(powerLimit);
+				if (powerLimit != null) {
+					this.log.warn("Setting PV limit: " + powerLimit + "W for " + this.pvInverterId);
+				} else {
+					this.log.info("No limit for " + this.pvInverterId);
+				}
+
+			}
+		} catch (OpenemsNamedException e) {
+			this.log.error("Error setting PV limit", e);
+		}
+
+	}
+
+	private Boolean getChannelValue(ChannelAddress address) throws OpenemsNamedException {
+		if (address == null) {
+			return null;
+		}
+		try {
+			BooleanReadChannel channel = this.componentManager.getChannel(address);
+			return channel.value().orElse(null);
+		} catch (OpenemsNamedException e) {
+			this.log.error("Error reading channel value", e);
+			return null;
+		}
+	}
+
+	@Override
+	public void run() throws OpenemsNamedException {
```

Comment:
Could be simplyfied with:

```java
private void setPvLimitBasedOnChannelValues(Boolean value0Percent, Boolean value30Percent, Boolean value60Percent, Boolean value100Percent) {
    if (value0Percent) {
        setPvLimit(0);
    } else if (value30Percent && !value60Percent && !value100Percent) {
        setPvLimit(this.powerLimit30Percent);
    } else if (value60Percent && !value30Percent && !value100Percent) {
        setPvLimit(this.powerLimit60Percent);
    } else if (value100Percent) {
        setPvLimit(null);
    } else {
        setPvLimit(0);
    }
} 
```

Reply:
I think it´s important to handle not (yet) exististing modbus channels. Otherwise limitation kicks in at every restart. The Exception in run() and the early return of the null-checks offers more safety. What do you think?